relayer/pkg/starknet/feeder_client.go

```diff
@@ -0,0 +1,211 @@
+	c.backoff = b
+	return c
+}
+
+func (c *FeederClient) WithMaxRetries(num int) *FeederClient {
+	c.maxRetries = num
+	return c
+}
+
+func (c *FeederClient) WithMaxWait(d time.Duration) *FeederClient {
+	c.maxWait = d
+	return c
+}
+
+func (c *FeederClient) WithMinWait(d time.Duration) *FeederClient {
+	c.minWait = d
+	return c
+}
+
+func (c *FeederClient) WithLogger(log utils.SimpleLogger) *FeederClient {
+	c.log = log
+	return c
+}
+
+func ExponentialBackoff(wait time.Duration) time.Duration {
+	return wait * 2
+}
+
+func NopBackoff(d time.Duration) time.Duration {
+	return 0
+}
+
+func NewFeederClient(clientURL string) *FeederClient {
+	return &FeederClient{
+		url:        clientURL,
+		client:     http.DefaultClient,
+		backoff:    ExponentialBackoff,
+		maxRetries: 5,
+		maxWait:    10 * time.Second,
+		minWait:    time.Second,
+		log:        utils.NewNopZapLogger(),
+	}
+}
+
+// buildQueryString builds the query url with encoded parameters
+func (c *FeederClient) buildQueryString(endpoint string, args map[string]string) string {
+	base, err := url.Parse(c.url)
+	if err != nil {
+		panic("Malformed feeder base URL")
+	}
+
+	base.Path += endpoint
+
+	params := url.Values{}
+	for k, v := range args {
+		params.Add(k, v)
+	}
+	base.RawQuery = params.Encode()
+
+	return base.String()
+}
+
+// get performs a "GET" http request with the given URL and returns the response body
+func (c *FeederClient) get(ctx context.Context, queryURL string) (io.ReadCloser, error) {
+	var res *http.Response
+	var err error
+	wait := time.Duration(0)
+	for i := 0; i <= c.maxRetries; i++ {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-time.After(wait):
+			var req *http.Request
+			req, err = http.NewRequestWithContext(ctx, "GET", queryURL, http.NoBody)
+			if err != nil {
+				return nil, err
+			}
+
+			res, err = c.client.Do(req)
+			if err == nil {
+				if res.StatusCode == http.StatusOK {
+					return res.Body, nil
+				}
+
+				if res.StatusCode != http.StatusOK {
+					err = errors.New(res.Status)
+				}
+				res.Body.Close()
+			}
+
+			if wait < c.minWait {
+				wait = c.minWait
+			}
+			wait = c.backoff(wait)
+			if wait > c.maxWait {
+				wait = c.maxWait
+			}
+			c.log.Warnw("failed query to feeder, retrying...", "retryAfter", wait.String())
+		}
+	}
+	return nil, err
+}
+
+func (c *FeederClient) TransactionFailure(ctx context.Context, transactionHash *felt.Felt) (*TransactionFailureReason, error) {
+	queryURL := c.buildQueryString("get_transaction", map[string]string{
+		"transactionHash": transactionHash.String(),
+	})
+
+	body, err := c.get(ctx, queryURL)
+
+	if err != nil {
+		return nil, err
+	}
+	defer body.Close()
+
+	txStatus := new(TransactionFailureReason)
+	if err = json.NewDecoder(body).Decode(txStatus); err != nil {
+		return nil, err
+	}
+	return txStatus, nil
+}
+
+// Only responds on valid /get_transaction?transactionHash=<TRANSACTION_HASH> requests. fails otherwise
+func NewTestServer() *httptest.Server {
```

Comment:
yeah, it started out in the feeder_client_test file. but then i had to put it here because i needed to import it in the integration pkg for integration tests. Go doesn't import exported variables from *_test.go files outside of the package.